src/relax/transform/gradient.cc

```diff
@@ -0,0 +1,405 @@
+    if (v->IsInstance<DataflowVarNode>()) {
+      builder_->Emit(VarBinding(v, e));
+    } else {
+      builder_->EmitOutput(VarBinding(v, e));
+    }
+  }
+
+  void CheckTarget(const Expr& e) {
+    ICHECK(!e->IsInstance<DataflowVarNode>()) << "not an output node";
+    ICHECK(e->checked_type().as<DynTensorTypeNode>()) << "target must be a DynTensorType";
+    ICHECK(e->shape().as<ShapeExprNode>()) << "error when getting target shape";
+    const auto* shape_node = e->shape().as<ShapeExprNode>();
+    ICHECK(shape_node->values.size() == 0) << "target must be a scalar";
+  }
+
+  void InitGrad(const Var& adjoint_var, const Var& var) {
+    ObjectPtr<InitAttrs> attrs = make_object<InitAttrs>();
+    auto type = Downcast<DynTensorType>(var->checked_type());
+    attrs->dtype = type->dtype;
+
+    const Expr& init = Call(ones_op_, {var->shape()}, Attrs(attrs));
+    adjoint_expr_map_.Set(var, init);
+  }
+
+  // the differentiation target
+  Var target_var_;
+  // the arguments to differentiate
+  Array<Var> require_grads_;
+
+  // var to its adjoints var
+  Map<Var, Var> adjoint_var_map_;
+  // var to its adjoint expr
+  Map<Var, Expr> adjoint_expr_map_;
+  // trace binding
+  Map<Expr, Var> adjoint_expr_to_var_;
+  // track zeros introduced
+  std::unordered_set<Expr, ObjectPtrHash, ObjectPtrEqual> zeros_tracker_;
+
+  // gop map
+  const OpAttrMap<FPrimalGradient> gradient_op_map_ =
+      Op::GetAttrMap<FPrimalGradient>("FPrimalGradient");
+
+  // constant
+  const Op& ones_op_ = Op::Get("relax.ones");
+  const Op& add_op_ = Op::Get("relax.add");
+  const Op& zeros_op_ = Op::Get("relax.zeros");
+};
+
+/*!
+ * \brief A simple reverse-mode auto differentiation.
+ * \param m The module
+ * \param var The GlobalVar of the specific function
+ * \param require_grad_names The relax variables which need adjoints. Must be inputs.
+ * \return The module after AD.
+ */
+IRModule Gradient(IRModule m, const GlobalVar& var, const Array<Var>& require_grads) {
+  BaseFunc base_func = m->Lookup(var);
+  if (auto* n = base_func.as<FunctionNode>()) {
+    auto f_before = GetRef<Function>(n);
+    Array<Var> require_grads_var;
+    for (auto input : require_grads) {
+      ICHECK(std::find(n->params.begin(), n->params.end(), input) != n->params.end())
+          << "function " << var->name_hint << " has no var named " << input->name_hint();
+      require_grads_var.push_back(input);
+    }
+
+    IRModuleNode* new_module_node = m.CopyOnWrite();
+    auto new_module = GetRef<IRModule>(new_module_node);
+    auto mutator = GradientMutator(new_module, require_grads_var);
+
+    auto adjoint_var = GlobalVar(var->name_hint + "_adjoint");
+    Function f_after = mutator.FuncTransform(f_before.as<FunctionNode>());
+    f_after = WithAttr(f_after, tvm::attr::kGlobalSymbol, adjoint_var->name_hint);
+    new_module->Add(adjoint_var, f_after);
+
+    return new_module;
+  } else {
+    LOG(FATAL) << "relax function " << var->name_hint << " not found";
+    return m;
+  }
```

Comment:
Usually we don’t want to have deep nesting of code blocks. One way is to write code like this
```c++
  auto* func = base_func.as<FunctionNode>();
  if (func == nullptr) {
    LOG(FATAL) << "...";
  }
  ...
```